mediatek/filogic: Make BY25Q128AS work with spi-nor-generic

themaverickdm opened the pull request:
BY25Q128AS in wl-wn586x3 was not being recognised by spi-nor-generic
with the error:
[    2.920569] spi-nor: probe of spi0.0 failed with error -5
Newer kernels use 16-bit status register writes by default
(spi_nor_write_16bit_cr_and_check) to set the quad mode, but this is
not supported by BY25Q128AS, so use dual mode (for now). And use 50
MHz bus speed - the original firmware uses 52 MHz actually.

Tested on wl-wn586x3.

themaverickdm commented:
@hauke, @Ansuel, @raenye, @musashino205, please review this hopefully small change. Thanks.

hauke commented:
Please report this to the upstream Linux mtd maintainer. Maybe they come up with a real solution for your problem.

themaverickdm commented:
@hauke, @Ansuel, @raenye, @musashino205, I apologise for spamming you all inadvertently with my previous comment!
I request you all to please review this small change, and merge it if it looks good.
Thanks.

themaverickdm commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -1247,7 +1247,7 @@ define Device/wavlink_wl-wn586x3
   DEVICE_DTS_LOADADDR := 0x47000000
   IMAGE_SIZE := 15424k
-  DEVICE_PACKAGES := kmod-mt7915e kmod-mt7981-firmware mt7981-wo-firmware
+  DEVICE_PACKAGES := kmod-mt76-connac kmod-mt7915e kmod-mt7981-firmware mt7981-wo-firmware
```
It is not.
Without this change, I get the following error if I try to build just the wl-wn586x3 target:
`make[3]: *** No rule to make target 'openwrt-w1/staging_dir/target-aarch64_cortex-a53_musl/usr/include/mac80211-backport/backport/autoconf.h', needed by 'openwrt-w1/build_dir/target-aarch64_cortex-a53_musl/linux-mediatek_filogic/mt76-2024.05.17~513c131c/.configured_b1c6ae92f8e974e60b47b3e5ad3573cc'.  Stop. `

themaverickdm replied:
I don't see the build failure anymore, so I have removed that change. See the new commit.

themaverickdm committed:
mediatek/filogic: Make BY25Q128AS work with spi-nor-generic

BY25Q128AS in wl-wn586x3 was not being recognised by spi-nor-generic
with the error:
[    2.920569] spi-nor: probe of spi0.0 failed with error -5
Newer kernels use 16-bit status register writes by default
(spi_nor_write_16bit_cr_and_check) to set the quad mode, but this is
not supported by BY25Q128AS, so use dual mode (for now). And use 50
MHz bus speed.

Tested on wl-wn586x3.

Signed-off-by: R Maru <deviantmaru@gmail.com>

981213 commented:
Hi!
The actual solution should be adding a new manufacturer file and a JEDEC ID entry to current spi-nor driver and apply the fixups there. Just like the macronix one you saw.

themaverickdm commented:
> Hi! The actual solution should be adding a new manufacturer file and a JEDEC ID entry to current spi-nor driver and apply the fixups there. Just like the macronix one you saw.

themaverickdm commented:
> Hi! The actual solution should be adding a new manufacturer file and a JEDEC ID entry to current spi-nor driver and apply the fixups there. Just like the macronix one you saw.

Thanks for your review, @981213.

In earlier kernels, yes, that was how it was done. Please see the patch under:
`target/linux/ramips/patches-6.6/405-mtd-spi-nor-Add-support-for-BoHong-bh25q128as.patch`

However, with newer kernels, we are trying to move away from manucfaturer-specific entries, and run purely based on SFDP data. And I have to say that it almost worked in quad mode, except that the 6.6.x spi-nor driver tries to use 16-bit status register writes by default, which are not supported by this flash part. So the best work-around for now is to use dual mode, which will atleast get openwrt working on this device, even with a lower data speed. In the meantime, we look for a way to get the quad mode working with SFDP.

981213 commented:
> In earlier kernels, yes, that was how it was done. Please see the patch under: `target/linux/ramips/patches-6.6/405-mtd-spi-nor-Add-support-for-BoHong-bh25q128as.patch`

Add the patch back with the flags you need then.
 
> However, with newer kernels, we are trying to move away from manucfaturer-specific entries, and run purely based on SFDP data. And I have to say that it almost worked in quad mode,

'Almost' isn't good enough. If a flash chip doesn't work correctly with the generic driver, it shouldn't be using it.
There are a lot of flash chips which don't provide enough info, or provide incorrect ones, in SFDP. And that's why the upstream still keep the id-based detection method.

> except that the 6.6.x spi-nor driver tries to use 16-bit status register writes by default (to set the Quad-Enable bit), which are not supported by this flash part. So the best work-around for now is to use dual mode, which will atleast get openwrt working on this device, even with a lower data speed. In the meantime, we look for a way to get the quad mode working with SFDP.
> 
> Please see my earlier comment for a possible solution (does not exist today though, requires upstream patches) which can tell the spi-nor driver to not use 16-bit status register writes: [#16601 (comment)](https://github.com/openwrt/openwrt/pull/16601#issuecomment-2394344439)

This approach will definitely get rejected. The linux kernel has moved away from specifying spi-nor chip models in device tree ages ago. Your proposal is just a different way of specifying that.

themaverickdm commented:
> Add the patch back with the flags you need then.

We should avoid this if the flash part works with the current spi-nor-generic, even if at a lower speed.

> 'Almost' isn't good enough. If a flash chip doesn't work correctly with the generic driver, it shouldn't be using that driver. There are a lot of flash chips which don't provide enough info, or provide incorrect ones, in SFDP. And that's why the upstream still keep the id-based detection method.

Honestly, I have to spend more time on this if/when I get a chance to figure out what can be done here. I know one thing for sure - it is better to have the device working, even if at a lower speed, rather than not having it work at all! And that too, with just a "tweaked device tree".

> This approach will definitely get rejected. The linux kernel has moved away from specifying spi-nor chip models in device tree ages ago. Your proposal is just a different way of specifying that.

Actually, we would not be adding entire make/models with this approach, only the bits (quite literally in this case!) necessary to make it work with spi-nor-generic. In this case, it is one of the flags in "spi_nor_option_flags"! In any case, this is far off on the horizon, and if I get to it. :) Let us get the device working first!

981213 commented:
> > Add the patch back with the flags you need then.
> 
> We should avoid this if the flash part works with the current spi-nor-generic, even if at a lower speed.

No. In your case SFDP data alone isn't providing enough info for a proper operation, and spi-nor-generic should not be used.
The id-based detection mechanism is kept for exactly this scenario. It makes no sense to avoid that in this case.

> Honestly, I have to spend more time on this if/when I get a chance to figure out what can be done here. I know one thing for sure - it is better to have the device working, even if at a lower speed, rather than not having it work at all! And that too, with just a "tweaked device tree".

With a simple kernel patch you can get the device working at full speed. It's unnecessary to compormise here.

> > This approach will definitely get rejected. The linux kernel has moved away from specifying spi-nor chip models in device tree ages ago. Your proposal is just a different way of specifying that.
> 
> Actually, we would not be adding entire make/models with this approach, only the bits (quite literally in this case!) necessary to make it work with spi-nor-generic. In this case, it is one of the flags in "spi_nor_option_flags"!

There aren't much difference between them. You are specifying that the flash onboard must have this BY25Q128AS-specific property. This limits the supported flash to this one, and maybe a handful models like it. Any other flash supporting SR2 won't work with this flag set.

> In any case, this is far off on the horizon, and if I get to it. :) Let us get the device working with openwrt first!

Just add a patch with JEDEC ID and required flags.

themaverickdm commented:
> > > This approach will definitely get rejected. The linux kernel has moved away from specifying spi-nor chip models in device tree ages ago. Your proposal is just a different way of specifying that.
> > 
> > 
> > Actually, we would not be adding entire make/models with this approach, only the bits (quite literally in this case!) necessary to make it work with spi-nor-generic. In this case, it is one of the flags in "spi_nor_option_flags"!
> 
> There aren't much difference between them. You are specifying that the flash onboard must have this BY25Q128AS-specific property. This limits the supported flash to this one, and maybe a handful models like it. Any other flash supporting SR2 won't work with this flag set.

This is a valid point. On the flip side, we don't even have to know what kind of flash chip is inside this device! :)

> > In any case, this is far off on the horizon, and if I get to it. :) Let us get the device working with openwrt first!
> 
> Just add a patch with JEDEC ID and required flags.

I tried moving:
`target/linux/ramips/patches-6.6/405-mtd-spi-nor-Add-support-for-BoHong-bh25q128as.patch`
into:
`target/linux/generic/pending-6.6/405-mtd-spi-nor-Add-support-for-BoHong-bh25q128as.patch
`
and I get the following:
```
[    2.909066] spi-nor spi0.0: CR: read back test failed
[    2.914121] spi-nor spi0.0: quad mode not supported
[    2.919014] spi-nor: probe of spi0.0 failed with error -5
```
I.e. even with the patch, this flash device is not working in quad mode. I think it's time to try another device.

981213 commented:
The patch you mentioned also missed the fixup function to unset SNOR_F_HAS_16BIT_SR. You need to add that function yourself.

dadogroove commented:
I have been playing around with this device lately to get it up and running and used this. Not sure if the FLAGS are correct but the device will boot and recognizes the chip as by25q128as.

`--- a/drivers/mtd/spi-nor/Makefile
+++ b/drivers/mtd/spi-nor/Makefile
@@ -2,6 +2,7 @@
 
 spi-nor-objs			:= core.o sfdp.o swp.o otp.o sysfs.o
 spi-nor-objs			+= atmel.o
+spi-nor-objs           += boya.o
 spi-nor-objs			+= catalyst.o
 spi-nor-objs			+= eon.o
 spi-nor-objs			+= esmt.o

--- /dev/null
+++ b/drivers/mtd/spi-nor/boya.c
@@ -0,0 +1,31 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (C) 2005, Intec Automation Inc.
+ * Copyright (C) 2014, Freescale Semiconductor, Inc.
+ */
+
+#include <linux/mtd/spi-nor.h>
+
+#include "core.h"
+
+static void by25q128as_default_init_fixups(struct spi_nor *nor)
+{
+	nor->flags &= ~SNOR_F_HAS_16BIT_SR;
+}
+
+static struct spi_nor_fixups by25q128as_fixups = {
+	.default_init = by25q128as_default_init_fixups,
+};
+
+static const struct flash_info boya_nor_parts[] = {
+       /* Boya Microelectronics */
+       { "by25q128as", INFO(0x684018, 0, 64 * 1024, 256)
+                FLAGS(SPI_NOR_HAS_LOCK | SPI_NOR_HAS_TB)
+                   .fixups = &by25q128as_fixups },
+};
+
+const struct spi_nor_manufacturer spi_nor_boya = {
+       .name = "boya",
+       .parts = boya_nor_parts,
+       .nparts = ARRAY_SIZE(boya_nor_parts),
+};

--- a/drivers/mtd/spi-nor/core.c
+++ b/drivers/mtd/spi-nor/core.c
@@ -2001,6 +2001,7 @@
 
 static const struct spi_nor_manufacturer *manufacturers[] = {
 	&spi_nor_atmel,
+   &spi_nor_boya,
 	&spi_nor_catalyst,
 	&spi_nor_eon,
 	&spi_nor_esmt,
--- a/drivers/mtd/spi-nor/core.h
+++ b/drivers/mtd/spi-nor/core.h
@@ -631,6 +631,7 @@
 
 /* Manufacturer drivers. */
 extern const struct spi_nor_manufacturer spi_nor_atmel;
+extern const struct spi_nor_manufacturer spi_nor_boya;
 extern const struct spi_nor_manufacturer spi_nor_catalyst;
 extern const struct spi_nor_manufacturer spi_nor_eon;
 extern const struct spi_nor_manufacturer spi_nor_esmt;`

dadogroove commented:
I have been trying to get this device working with OpenWrt and have had this problem with the Boya chip. I have been using this patch successfully. I am not sure that the FLAGS are correct but it will boot up and recognize the chip as by25q128as. 

[345-mtd-spi-nor-Add-support-for-Boya-by25q128as.PATCH](https://github.com/user-attachments/files/17441620/345-mtd-spi-nor-Add-support-for-Boya-by25q128as.PATCH)

On another note, the snapshot sysupgrade files for this device that are generated work for install but not for sysupgrades. The files generated are not tar files. It causes the router to panic and bootloop.

hauke commented:
@dadogroove the patch looks good. I prefer this approach to solve the problem. It should use tabs instead of spaces. Please include the sfdp data in the patch description too. It would be nice if you could send it upstream.

An other flash chip had the same problem and they fixed it like this upstream: https://github.com/torvalds/linux/commit/6823a8383420263bc061865027836755615a275f

themaverickdm commented:
Please see https://github.com/openwrt/openwrt/pull/17462 where I have tried to incorporate @dadogroove's patch mentioned above, with changes suggested by @hauke.
